src/monmove.cpp

```diff
@@ -1442,3 +1565,36 @@ int monster::turns_to_reach( int x, int y )
 
     return int( turns + .9 ); // Halve (to get turns) and round up
 }
+
+item_location get_most_desired_item_in_radius( tripoint pos, int radius,
+        std::function<bool( const tripoint & ) > tile_qualifies,
+        std::function<float( const item & )> calc_desirability )
+{
+    float max_desirability = -1;
+    tripoint desired_item_pos;
+    item *desired_item = nullptr;
+    for( auto &p : g->m.points_in_radius( pos, radius ) ) {
+        if( tile_qualifies ) {
```

Comment:
You can speed up the checks (`map::i_at` is relatively expensive) by checking if items exist before that.
The cost isn't high, but you'd be calling this over a giant area, which could easily be noticeable if large number of creatures do this.